eclass/rust.eclass

```diff
@@ -0,0 +1,480 @@
+		usedep=${1}
+	fi
+
+	local slot
+
+	local RUST_DEPS=()
+	RUST_DEPS+=( "|| (" )
+	for slot in "${_RUST_SLOTS[@]}"; do
+		if [[ -n ${usedep} ]]; then
+			RUST_DEPS+=( "dev-lang/rust:${slot}[${usedep}] dev-lang/rust-bin:${slot}[${usedep}]" )
+		else
+			RUST_DEPS+=( "dev-lang/rust:${slot} dev-lang/rust-bin:${slot}" )
+		fi
+	done
+	RUST_DEPS+=( ")" )
+	echo "${RUST_DEPS[@]}"
+}
+
+# == ebuild helpers ==
+
+# @FUNCTION: rust_check_usedep
+# @USAGE: [-b|-d] slot usedep
+# @DESCRIPTION:
+# Check if a Rust package with the specified SLOT and USE-dependency
+# is installed. If -b is specified, the checks are performed relative
+# to BROOT. If -d is specified, the checks are performed relative to
+# ESYSROOT. -d is the default.
+#
+# Returns SLOT and Rust type if the slot is suitable.
+# Rust type is either "source" or "binary".
+rust_check_usedep()
+{
+	local hv_switch=-d
+	while [[ ${1} == -* ]]; do
+		case ${1} in
+			-b|-d) hv_switch=${1};;
+			*) break;;
+		esac
+		shift
+	done
+
+	local slot=${1}
+	local usedep=${2}
+
+	if ( has_version ${hv_switch} "dev-lang/rust:${slot}[${usedep}]" ||
+		has_version ${hv_switch} "dev-lang/rust-bin:${slot}[${usedep}]" ); then
+		if has_version ${hv_switch} "dev-lang/rust:${slot}[${usedep}]"; then
+			rust_type="source"
+		else
+			rust_type="binary"
+		fi
+		echo ${slot}
+		echo ${rust_type}
+	fi
+}
+
+# @FUNCTION: get_rust_slot
+# @USAGE: [-b|-d]
+# @DESCRIPTION:
+# Find the newest Rust install that is acceptable for the package,
+# and print its version number (i.e. SLOT) and type (source or bin[ary]).
+#
+# If -b is specified, the checks are performed relative to BROOT,
+# and BROOT-path is returned.
+#
+# If -d is specified, the checks are performed relative to ESYSROOT,
+# and ESYSROOT-path is returned. -d is the default.
+#
+# If RUST_M{AX,IN}_SLOT is non-zero, then only Rust versions that
+# are not newer or older than the specified slot(s) will be considered.
+# Otherwise, all Rust versions are be considered acceptable.
+#
+# If the `rust_check_deps()` function is defined within the ebuild, it
+# will be called to verify whether a particular slot is accepable.
+# Within the function scope, RUST_SLOT and LLVM_SLOT will be defined.
+#
+# The function should return a true status if the slot is acceptable,
+# false otherwise. If rust_check_deps() is not defined, the function
+# defaults to checking whether a suitable Rust package is installed.
+get_rust_slot() {
+	debug-print-function ${FUNCNAME} "$@"
+
+	local hv_switch=-d
+	while [[ ${1} == -* ]]; do
+		case ${1} in
+			-b|-d) hv_switch=${1};;
+			*) break;;
+		esac
+		shift
+	done
+
+	local max_slot
+	if [[ -z ${RUST_MAX_VER} ]]; then
+		max_slot=
+	else
+		max_slot=${RUST_MAX_VER}
+	fi
+	local slot
+	local llvm_slot
+
+	if [[ -n ${RUST_NEEDS_LLVM} ]]; then
+		local llvm_r1_slot
+		# quickly get a list of unique llvm slots that we support
+		for llvm_slot in $(echo "${_RUST_KNOWN_SLOTS[@]}"  | tr ' ' '\n' | sort -u | tr '\n' ' '); do
```

Comment:
```suggestion
		for llvm_slot in $(echo "${_RUST_KNOWN_SLOTS[@]}" | tr ' ' '\n' | sort -u -n | tr '\n' ' '); do
```

I'm pretty sure you don't want lexical sorting for numbers. Also the fact that you've missed that kinda proves that this is too fragile.

Reply:
We care that we're getting a list of unique slots here than the actual order, but that's also fine.

```
numbers=(
 18
 18
 18
 17
 19
 20
)
$ echo ${numbers[@]}
18 18 18 17 19 20
$ echo "${numbers[@]}"  | tr ' ' '\n' | sort -u | tr '\n' ' '
17 18 19 20
```

Reply:
The keys of an associative array are unique.